src/riak_kv_qry_buffers.erl

```diff
@@ -660,6 +743,57 @@ tstamp() ->
 
 
 %% data ops
+
+-spec enkey([data_row()],
+            KeyFieldPositions::[non_neg_integer()],
+            OrdByFieldQualifiers::[{asc|desc, nulls_first|nulls_last}],
+            ChunkId::non_neg_integer()) ->
+                   [{{KeyOrd::riak_pb_ts_codec:ldbvalue(),
+                      ChunkId::non_neg_integer(),
+                      Idx::non_neg_integer()},
+                     data_row()}].
+enkey(Rows, KeyFieldPositions, OrdByFieldQualifiers, ChunkId)
+  when is_list(KeyFieldPositions) ->
+    %% 0. The new key is composed from fields appearing in the ORDER
+    %%    BY clause, and may therefore not work out to be unique. We
+    %%    now index the rows in the chunk to preserve the original
+    %%    order (because imposing any other order is even worse)
+    RowsIndexed = lists:zip(Rows, lists:seq(1, length(Rows))),
+    [begin
+         %% a. Form a new key from ORDER BY fields
+         KeyRaw = [lists:nth(Pos, Row) || Pos <- KeyFieldPositions],
+         %% b. Negate values in columns marked as DESC in ORDER BY clause
+         KeyOrd = [make_sorted_keys(F, AscDesc, Nulls)
+                   || {F, {AscDesc, Nulls}} <- lists:zip(KeyRaw, OrdByFieldQualifiers)],
+         %% c. Combine with chunk id and row idx to ensure uniqueness, and encode.
+         %%    Make sure the key proper goes first: ChunkId is a thing
+         %%    internal to query buffer, is incrementally increasing,
+         %%    whereas the chunk it refers to may certainly not be
+         %%    increasing because of out-of-order arrivals from the
+         %%    vnodes.
+         Key = {KeyOrd, ChunkId, Idx},
+         %% d. Encode the record (don't bother constructing a
+         %%    riak_object with metadata, vclocks):
+         {Key, Row}
+     end || {Row, Idx} <- RowsIndexed].
+
+make_sorted_keys([], desc, nulls_first) ->
```

Comment:
The first 4 function heads can be collapsed to two; `nulls_first` (or `_last`) is the only parameter that matters on an empty list.

Having said that, I'm also curious whether that's the intent. It feels like `nulls_last` should put them first if also `desc`.

Please either collapse the heads or fix the logic.